go/vt/vttablet/grpctmclient/cached_client.go

```diff
@@ -0,0 +1,340 @@
+	*queue = append(*queue, conn)
+}
+
+// Pop is part of the container/heap.Interface interface.
+func (queue *cachedConns) Pop() interface{} {
+	old := *queue
+	n := len(old)
+	conn := old[n-1]
+	old[n-1] = nil  // avoid memory leak
+	conn.index = -1 // for safety
+	*queue = old[0 : n-1]
+
+	return conn
+}
+
+type cachedConnDialer struct {
+	m            sync.RWMutex
+	conns        map[string]*cachedConn
+	qMu          sync.Mutex
+	queue        cachedConns
+	connWaitSema *sync2.Semaphore
+}
+
+var dialerStats = struct {
+	ConnReuse            *stats.Gauge
+	ConnNew              *stats.Gauge
+	DialTimeouts         *stats.Gauge
+	DialTimings          *stats.Timings
+	EvictionQueueTimings *stats.Timings
+}{
+	ConnReuse:    stats.NewGauge("tabletmanagerclient_cachedconn_reuse", "number of times a call to dial() was able to reuse an existing connection"),
+	ConnNew:      stats.NewGauge("tabletmanagerclient_cachedconn_new", "number of times a call to dial() resulted in a dialing a new grpc clientconn"),
+	DialTimeouts: stats.NewGauge("tabletmanagerclient_cachedconn_dial_timeouts", "number of context timeouts during dial()"),
+	DialTimings:  stats.NewTimings("tabletmanagerclient_cachedconn_dialtimings", "timings for various dial paths", "path", "rlock_fast", "sema_fast", "sema_poll"),
+	// TODO: add timings for heap operations (push, pop, fix)
+}
+
+// NewCachedConnClient returns a grpc Client using the priority queue cache
+// dialer implementation.
+func NewCachedConnClient(capacity int) *Client {
+	dialer := &cachedConnDialer{
+		conns:        make(map[string]*cachedConn, capacity),
+		queue:        make(cachedConns, 0, capacity),
+		connWaitSema: sync2.NewSemaphore(capacity, 0),
+	}
+
+	heap.Init(&dialer.queue)
+	return &Client{dialer}
+}
+
+var _ dialer = (*cachedConnDialer)(nil)
+
+func (dialer *cachedConnDialer) dial(ctx context.Context, tablet *topodatapb.Tablet) (tabletmanagerservicepb.TabletManagerClient, io.Closer, error) {
+	start := time.Now()
+
+	addr := getTabletAddr(tablet)
+	dialer.m.RLock()
+	if conn, ok := dialer.conns[addr]; ok {
+		defer func() {
+			dialerStats.DialTimings.Add("rlock_fast", time.Since(start))
+		}()
+		defer dialer.m.RUnlock()
+		return dialer.redial(conn)
+	}
+	dialer.m.RUnlock()
+
+	if dialer.connWaitSema.TryAcquire() {
+		defer func() {
+			dialerStats.DialTimings.Add("sema_fast", time.Since(start))
+		}()
+		dialer.m.Lock()
+		defer dialer.m.Unlock()
+
+		// Check if another goroutine managed to dial a conn for the same addr
+		// while we were waiting for the write lock. This is identical to the
+		// read-lock section above.
+		if conn, ok := dialer.conns[addr]; ok {
+			return dialer.redial(conn)
+		}
+
+		return dialer.newdial(addr, true /* manage queue lock */)
+	}
+
+	defer func() {
+		dialerStats.DialTimings.Add("sema_poll", time.Since(start))
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			dialerStats.DialTimeouts.Add(1)
+			return nil, nil, ctx.Err()
+		default:
+			dialer.m.Lock()
```

Comment:
same function decomposition for locking thing here. If you can make it a separate function, most of the unlocks can just be immediate deferrals, instead of having to manage them around the for loop